core/psh/psh.c

```diff
@@ -972,7 +975,113 @@ static void psh_signalstop(int sig)
 }
 
 
-static int psh_run(void)
+static int psh_logincredentget(char *buff, int ispasswd, int maxinlen)
+{
+	int inputlen = -1;
+	char c;
+
+	for (;;) {
+		if (inputlen < 0) {
+			if (ispasswd)
+				fprintf(stdout, "Password: ");
+			else
+				fprintf(stdout, "Login: ");
+
+			fflush(stdout);
+			inputlen++;
+		}
+
+		if (read(STDIN_FILENO, &c, 1) != 1)
+			continue;
+			
+		if (c == '\n') {
+			if (inputlen <= maxinlen && inputlen > 0) {
+				buff[inputlen] = '\0';
+				return --inputlen;
+			}
+			else {
+				inputlen = -1;
+				if (ispasswd) {
+					memset(buff, '\0', sizeof(char)*maxinlen);
+					return 0;
+				}
+				else
+					continue;
+			}
+		}
+		else if (isprint(c)) {
+			if (inputlen < maxinlen)
+				buff[inputlen] = c;
+			inputlen++;
+			continue;
+		}
+	}
+	return -EINTR;
+}
+
+
+static int psh_login(void)
+{
+	int err = 0, auth = -EACCES;
+	const int maxlen = 32;
+	char username[maxlen+1], passwd[maxlen+1], *shadow;
+	struct passwd *userdata;
+	struct termios origattr, ttyattr;
+
+	/* get/prepare tty settings */
+	if ((err = tcgetattr(STDIN_FILENO, &origattr)) < 0 || (err = tcgetattr(STDIN_FILENO, &ttyattr)) < 0) {
```

Comment:
You are right, calling tcgetattr twice is redundant. Will be changed to one call and structure-to-structure assignment. `err` is clearly dead code - removed. 

I thought about moving `tcgetattr`/`tcsetattr` to `psh_logincredentget` but:
- it would again increase the number of tty interactions (1 get and 2 sets for each subroutine call) 
- if not above then we can `tcgetattr` once in psh_login and pass it to the subroutine (unnecessary code fragmentation)
- there is a problem with `Login failed` which needs to be presented in echo mode, but presented based on login logic outcome (removing this prompt doesn`t seem a big loss so at least its not a big problem)

I think that the best solution is to trade one more tty interaction for better code readability and move tty setters and getters to subroutine and remove `Login failed` prompt completely as it is not very crucial feature.

Reply:
> it would again increase the number of tty interactions (1 get and 2 sets for each subroutine call)

I think we may not care, user input would be slow regardless

> there is a problem with Login failed which needs to be presented in echo mode, but presented based on login logic outcome (removing this prompt doesn`t seem a big loss so at least its not a big problem)

I think it should work (we restore the original TTY settings after password input, so it should be visible). Experiments on host confirm my suspicion, so if it doesn't work on our OS the same way - it needs to be fixed (note: maybe it's because of using C streams - adding explicit `fflush()` might help).
![Screenshot_20210415_132758](https://user-images.githubusercontent.com/3939983/114862246-93a87480-9dee-11eb-9ffc-cadea5994f2b.png)


Side note: on host I can exit from first pshlogin by sending signals via tty (`^C`, `^Z`, '^\`), but the second login (after exiting from PSH) cannot be stopped the same way, maybe the login prompt could be strengthened:

![Screenshot_20210415_133158](https://user-images.githubusercontent.com/3939983/114862561-f7cb3880-9dee-11eb-952d-1aad08fe9fdd.png)

The above problems can be fixed as a separate PR, let's try to close this one ASAP :)